lib/API/api.cpp

```diff
@@ -647,23 +718,39 @@ llvm::Error compile_(int argc, char const **argv, std::string *outputString,
                                   commandLinePassesTiming))
     return err;
 
-  if (pm.size() && failed(pm.run(moduleOp)))
+  if (pm.size() && failed(pm.run(moduleOp))) {
+    emitAllDiagnosticsAndCheckForErrors_(target, diagnosticCb, config);
     return llvm::createStringError(llvm::inconvertibleErrorCode(),
                                    "Problems running the compiler pipeline!");
+  }
   commandLinePassesTiming.stop();
 
+  if (emitAllDiagnosticsAndCheckForErrors_(target, diagnosticCb, config)) {
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "Problems running the compiler pipeline!");
+  }
+
   // Prepare outputs
   if (config.getEmitAction() == EmitAction::MLIR) {
     if (auto err = emitMLIR_(ostream, context, moduleOp, config,
-                             targetCompilationManager, errorHandler, timing))
+                             targetCompilationManager, errorHandler, timing)) {
+      emitAllDiagnosticsAndCheckForErrors_(target, diagnosticCb, config);
       return err;
+    }
   }
 
   if (config.getEmitAction() == EmitAction::QEM ||
       config.getEmitAction() == EmitAction::QEQEM) {
     if (auto err = emitQEM_(config, ostream, std::move(payload), moduleOp,
-                            targetCompilationManager, timing))
+                            targetCompilationManager, timing)) {
+      emitAllDiagnosticsAndCheckForErrors_(target, diagnosticCb, config);
       return err;
+    }
+  }
+
+  if (emitAllDiagnosticsAndCheckForErrors_(target, diagnosticCb, config)) {
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "Problems generating payload data!");
   }
```

Comment:
Instead of separating out checking for diagnostics, I believe we should incorporate this directly into the respective target methods (Ie., the ThreadedCompilationManager or target initialization) to localize the detection of errors to where the targets are invoked)

Reply:
My previous failed solution involved sending the callback to the targets during initialization, so I'm quite wary of returning to that. I'll look into adding it to the compilation manager processing to see if that's cleaner.

Reply:
I agree that this is not so nice currently, but it does work at least. 😝

Reply:
I think it should be possible since we're not invoking the callback per-se, we should just be checking if there are errors and accumulating them in the compilation manager

Reply:
Removed all of this with things moved to the TargetCompilationManager.